### PR TITLE
Fuzz v2

### PR DESCRIPTION
**Pull request by catenacyber**
Link to redmine ticket:
https://redmine.openinfosecfoundation.org/issues/2859

Describe changes:
- Adds two fuzz targets
- Change compilation scheme so as to compile fuzz targets (option --enable-fuzztargets for configure)

Modifies #3813 
- renames fuzz targets to match the functions names
- create a new file init.c for common initialization between real suricata and fuzz targets (that means getting code out of suricata.c into init.c)
- calls `ConfInit` before `WindowsInitService` 
- new function `initGlobal` to do all the initialization before command line parsing

I think that this PR is not complete but I would like some feedback, that is why I am doing it.
I kept the history for commits so as to follow my logic.

There remains some more initialization logic in the fuzz targets.
These are inspired by afl-fuzz logic from `ParseCommandLineAFL` in suricata.c
Should they be merged together into one common initialization function ?
Do you know if these steps (like `MpmTableSetup`) depend on command line arguments ? 
(so should we merge them either into `initGlobal` or into some new `initWithArgs`)

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -2369,6 +2369,21 @@ fi
...
+    AC_CHECK_LIB(FuzzingEngine, main,, LIBFUZZINGENGINE="no")
+    if test "$LIBFUZZINGENGINE" = "no"; then
+        libfuzzingengine=no
+    fi
+    AM_CONDITIONAL([HAVE_LIBFUZZINGENGINE], [test "x$libfuzzingengine" != "xno"])
+
+    AC_ARG_ENABLE(fuzztargets,
+        AS_HELP_STRING([--enable-fuzztargets], [Enable fuzz targets]),[enable_fuzztargets=$enableval],[enable_fuzztargets=no])
+    AS_IF([test "x$enable_fuzztargets" = "xyes"], [
+        echo "   lolfuzz required"
```
Comment (victorjulien) [inline configure.ac]: lolfuzz?

File: src/init.c [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+int SuriHasSigFile(void)
+{
+    return (suricata.sig_file != NULL);
+}
+
+SuricataContext context;
+
+// Global initialization common to all runmodes
+int initGlobal() {
```
Comment (victorjulien) [inline src/init.c]: style: function names start with capital

File: src/tests/fuzz/fuzz_applayerparserparse.c [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+#include "app-layer-parser.h"
+
+#define HEADER_LEN 6
+
+void fuzz_openFile(const char * name) {
+}
+
+AppLayerParserThreadCtx *alp_tctx = NULL;
+
+int LLVMFuzzerTestOneInput(const uint8_t *Data, size_t Size) {
```
Comment (victorjulien) [inline src/tests/fuzz/fuzz_applayerparserparse.c]: style: variable & argument names should be all lowercase

File: src/tests/fuzz/fuzz_applayerparserparse.c [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+
+#include "suricata-common.h"
+#include "app-layer-detect-proto.h"
+#include "flow-util.h"
+#include "app-layer-parser.h"
+
+#define HEADER_LEN 6
+
+void fuzz_openFile(const char * name) {
```
Comment (victorjulien) [inline src/tests/fuzz/fuzz_applayerparserparse.c]: what is this for?

File: src/tests/fuzz/fuzz_applayerparserparse.c [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        AppLayerProtoDetectSetup();
+        AppLayerParserSetup();
+        AppLayerParserRegisterProtocolParsers();
+        alp_tctx = AppLayerParserThreadCtxAlloc();
+    }
+
+    if (Data[0] >= ALPROTO_MAX) {
+        return 0;
+    }
+    f = FlowAlloc();
```
Comment (victorjulien) [inline src/tests/fuzz/fuzz_applayerparserparse.c]: error check?

**Comment by catenacyber** [comment]
> Could you add a howto document to the PR in the next iteration? I'd like to try out the code.

Do you mean to add some file src/tests/fuzz/README with the content of
https://github.com/OISF/suricata/pull/3813#issuecomment-486936174 ?

File: configure.ac (hunk 1 above)
Comment (catenacyber) [reply configure.ac]: ok lol I forgot to remove this test

File: src/init.c (hunk 2 above)
Comment (catenacyber) [reply src/init.c]: ok

**Comment by victorjulien** [comment]
Yeah, that would be a good start.

File: src/tests/fuzz/fuzz_applayerparserparse.c (hunk 4 above)
Comment (catenacyber) [reply src/tests/fuzz/fuzz_applayerparserparse.c]: That is for onefile.c, a simple driver for which you can specify a output file
Let's wait a bit to see if we need it

File: src/tests/fuzz/fuzz_applayerparserparse.c (hunk 3 above)
Comment (catenacyber) [reply src/tests/fuzz/fuzz_applayerparserparse.c]: Ok, I was using libfuzzer style https://llvm.org/docs/LibFuzzer.html#id22

File: src/tests/fuzz/fuzz_applayerparserparse.c (hunk 5 above)
Comment (catenacyber) [reply src/tests/fuzz/fuzz_applayerparserparse.c]: ok

**Pull request closed without merging**